multicluster/controllers/multicluster/core/remote_cluster_manager_test.go

```diff
@@ -0,0 +1,146 @@
+
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
+
+	mcsv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+)
+
+var (
+	rcmtScheme = runtime.NewScheme()
+)
+
+func TestAddMember(t *testing.T) {
+
+	remoteClusterMgrUnderTest := NewRemoteClusterManager(common.ClusterSetID("clusterSetA"), common.ClusterID("memberA"))
+
+	remoteClusterMgrUnderTest.Start()
+
+	mockCtrl := gomock.NewController(t)
+	mockRemoteCluster := NewMockRemoteCluster(mockCtrl)
+
+	mockRemoteCluster.EXPECT().GetClusterID().Return(common.ClusterID("leaderA")).AnyTimes()
+	mockRemoteCluster.EXPECT().IsConnected().Return(true).AnyTimes()
+	mockRemoteCluster.EXPECT().Start()
+	mockRemoteCluster.EXPECT().StartMonitoring()
+
+	remoteClusterMgrUnderTest.AddRemoteCluster(mockRemoteCluster)
+
+	done := make(chan bool)
+
+	go func() {
+
+		for i := 0; i < 20; i++ {
+			time.Sleep(2 * time.Second)
+
+			if remoteClusterMgrUnderTest.GetElectedLeaderClusterID() != "leaderA" {
+				klog.Infof("Leader election not complete yet")
+				continue
+			}
+			klog.Infof("Leader election completed")
+
+			done <- true
+		}
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(20 * time.Second):
+		panic("timeout")
+	}
+
+	remoteClusterMgrUnderTest.Stop()
+}
+
+func TestLeaderElection(t *testing.T) {
+
+	remoteClusterMgrUnderTest := NewRemoteClusterManager("clusterSetA", "memberA")
+
+	remoteClusterMgrUnderTest.Start()
+
+	remoteClusterMgrUnderTest.Start()
+
+	mockCtrl := gomock.NewController(t)
+	mockRemoteCluster1 := NewMockRemoteCluster(mockCtrl)
+	mockRemoteCluster1.EXPECT().GetClusterID().Return(common.ClusterID("leaderA1")).AnyTimes()
+	mockRemoteCluster1.EXPECT().IsConnected().Return(true).Return(true).
+		Return(false).Return(false).Return(false)
+	mockRemoteCluster1.EXPECT().Start()
+	mockRemoteCluster1.EXPECT().StartMonitoring()
+
+	mockRemoteCluster2 := NewMockRemoteCluster(mockCtrl)
+	mockRemoteCluster2.EXPECT().GetClusterID().Return(common.ClusterID("leaderA2")).AnyTimes()
+	mockRemoteCluster2.EXPECT().IsConnected().Return(false).Return(false).
+		Return(true).Return(true).Return(true)
+	mockRemoteCluster2.EXPECT().Start()
+	mockRemoteCluster2.EXPECT().StartMonitoring()
+
+	remoteClusterMgrUnderTest.AddRemoteCluster(mockRemoteCluster1)
+	remoteClusterMgrUnderTest.AddRemoteCluster(mockRemoteCluster2)
+
+	done := make(chan bool)
+
+	go func() {
+
+		firstElectionDone := false
+		for i := 0; i < 20; i++ {
+			time.Sleep(2 * time.Second)
+			if !firstElectionDone {
+				leader := remoteClusterMgrUnderTest.GetElectedLeaderClusterID()
+				if leader != "leaderA1" {
+					klog.Infof("Leader election not complete yet")
+					continue
+				}
+				klog.Infof("Leader election completed")
+				firstElectionDone = true
+			}
+
+			if remoteClusterMgrUnderTest.GetElectedLeaderClusterID() != "leaderA2" {
+				klog.Infof("Leader has not changed yet")
+				continue
+			}
+
+			done <- true
+		}
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(25 * time.Second):
```

Comment:
you have 20 times retry and 2s sleep, but give it only 25s total timeout, is this expected?  btw, I tried rebase and looks like this test case failed with timeout even after I increase the 25 to 50s, Could you help to double check this case? thanks.

Reply:
here is what I got 
```
--- FAIL: TestLeaderElection (50.00s)
    leader_elector.go:111: Unexpected call to *core.MockRemoteCluster.IsConnected([]) at /Users/luolan/workspaces/codes/src/github.com/luolanzone/antrea/multicluster/controllers/multicluster/core/leader_elector.go:111 because:
        expected call at /Users/luolan/workspaces/codes/src/github.com/luolanzone/antrea/multicluster/controllers/multicluster/core/remote_cluster_manager_test.go:88 has already been called the max number of times
    controller.go:137: missing call(s) to *core.MockRemoteCluster.StartMonitoring() /Users/luolan/workspaces/codes/src/github.com/luolanzone/antrea/multicluster/controllers/multicluster/core/remote_cluster_manager_test.go:91
    controller.go:137: missing call(s) to *core.MockRemoteCluster.StartMonitoring() /Users/luolan/workspaces/codes/src/github.com/luolanzone/antrea/multicluster/controllers/multicluster/core/remote_cluster_manager_test.go:98
    controller.go:137: aborting test due to missing call(s)
panic: timeout [recovered]
	panic: timeout
```

Reply:
Yes, I am seeing that also, will debug.